x/staking/keeper/msg_server.go

```diff
@@ -605,6 +608,83 @@ func (k msgServer) UpdateParams(ctx context.Context, msg *types.MsgUpdateParams)
 	return &types.MsgUpdateParamsResponse{}, nil
 }
 
-func (k msgServer) RotateConsPubKey(_ context.Context, _ *types.MsgRotateConsPubKey) (*types.MsgRotateConsPubKeyResponse, error) {
-	return nil, nil
+func (k msgServer) RotateConsPubKey(goCtx context.Context, msg *types.MsgRotateConsPubKey) (res *types.MsgRotateConsPubKeyResponse, err error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	cv := msg.NewPubkey.GetCachedValue()
+	pk, ok := cv.(cryptotypes.PubKey)
+	if !ok {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidType, "expecting cryptotypes.PubKey, got %T", cv)
+	}
+
+	_, err = k.RotatedConsKeyMapIndex.Get(ctx, pk.Address())
+	if err != nil && !errors.Is(err, collections.ErrNotFound) {
```

Comment:
This currently throws out all generic errors, but you defined an error for when the address is already in use. Could we please use that in concrete comparison and also return it down below? Would be useful to also test against that condition to lock that behaviour in.

Reply:
This code is not for the address already in use, this is added to handle the previously rotated keys cannot be migrated back. (for ex: A->B is rotated and later we cannot rotate back to the key A)
This logic seems wrong though, need modifications.